dll/ntdll/ldr/appcompat.c

```diff
@@ -0,0 +1,386 @@
+        OrdinalTable[j] = OrgOrdinalTable[i];
+    }
+
+    /* Free the copy */
+    RtlFreeHeap(RtlGetProcessHeap(), 0, OrgNameTable);
+
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+LdrpPatchExportTable(
+    _In_ PVOID DllBase,
+    _In_ PROSCOMPAT_DESCRIPTOR RosCompatDescriptor,
+    _In_ DWORD AppCompatVersion
+    )
+{
+    PIMAGE_EXPORT_DIRECTORY ExportDirectory, MagicExportDir;
+    ULONG ExportDirectorySize, OldProtect;
+    PVOID ProtectAddress;
+    SIZE_T ProtectSize;
+    PULONG FunctionTable;
+    NTSTATUS Status;
+
+    //__debugbreak();
+
+    /* Get export directory */
+    ExportDirectory = RtlImageDirectoryEntryToData(DllBase,
+                                                   TRUE,
+                                                   IMAGE_DIRECTORY_ENTRY_EXPORT,
+                                                   &ExportDirectorySize);
+    if (ExportDirectory == NULL)
+    {
+        return STATUS_INVALID_IMAGE_FORMAT;
+    }
+
+    ASSERT(RosCompatDescriptor->NumberOfExportNames == ExportDirectory->NumberOfNames);
+
+    /* Unprotect the export directory */
+    ProtectAddress = ExportDirectory;
+    ProtectSize = ExportDirectorySize;
+    Status = NtProtectVirtualMemory(NtCurrentProcess(),
+                                    &ProtectAddress,
+                                    &ProtectSize,
+                                    PAGE_READWRITE,
+                                    &OldProtect);
+    if (!NT_SUCCESS(Status))
+    {
+        /* Fail */
+        //DbgPrint("LDR: Unable to unprotect IAT for %wZ (Status %x)\n",
+        //         &ImportLdrEntry->BaseDllName,
+        //         Status);
+        return Status;
+    }
+
+    /* Patch the export name and ordinal table */
+    Status = LdrpPatchExportNameTable(DllBase,
+                                      RosCompatDescriptor,
+                                      ExportDirectory,
+                                      AppCompatVersion);
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    FunctionTable = (PULONG)((ULONG_PTR)DllBase +
+        (ULONG_PTR)ExportDirectory->AddressOfFunctions);
+
+    /* Unprotect the export directory */
+    ProtectSize = ExportDirectorySize;
+    Status = NtProtectVirtualMemory(NtCurrentProcess(),
+                                    (PVOID*)&ExportDirectory,
+                                    &ProtectSize,
+                                    OldProtect,
+                                    &OldProtect);
+    if (!NT_SUCCESS(Status))
+    {
+        /* Fail */
+        //DbgPrint("LDR: Unable to unprotect export table for %wZ (Status %x)\n",
+        //         &ImportLdrEntry->BaseDllName,
+        //         Status);
+        return Status;
+    }
+
+    /* Also flush out the cache */
+    NtFlushInstructionCache(NtCurrentProcess(), ExportDirectory, ProtectSize);
+
+    /* Setup an export directory for "magic" (i.e. private) exports */
+    MagicExportDir = RosCompatDescriptor->MagicExportDir;
```

Comment:
`ASSERT(MagicExportDir);`

Reply:
I'm thinking of removing this part, because currently it's not used. I added it and then noticed that this doesn't work well together with our ldr code.